src/wrap/message.rs

```diff
@@ -62,3 +80,96 @@ impl Into<Vec<u8>> for Message {
         vec
     }
 }
+
+impl Message {
+    /// Decrypt the [`Message`] with the provided [`wrap::Key`]
+    pub fn decrypt(&self, key: &wrap::Key) -> Result<Plaintext, Error> {
+        let cipher: super::key::AesCcm = key.into();
+        let plaintext = cipher
+            .decrypt(&self.nonce.to_nonce(), &*self.ciphertext)
+            .unwrap();
+
+        let plaintext = deserialize(&plaintext).unwrap();
+        Ok(plaintext)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Plaintext {
+    pub alg_id: Algorithm,
+    pub object_info: wrap::Info,
+    pub data: Vec<u8>,
+}
+
+impl Plaintext {
+    pub fn encrypt(&self, key: &super::Key) -> Result<Message, Error> {
+        let cipher: super::key::AesCcm = key.into();
+        let nonce = Nonce::generate();
+        let wire = serialize(&self).unwrap();
+        let ciphertext = cipher.encrypt(&nonce.to_nonce(), wire.as_slice()).unwrap();
+
+        Ok(Message { nonce, ciphertext })
+    }
+
+    pub fn ecdsa<C>(&self) -> Option<SecretKey<C>>
+    where
+        C: PrimeCurve + CurveAlgorithm + ValidatePublicKey,
+        FieldBytesSize<C>: ModulusSize + Unsigned,
+    {
+        if let algorithm::Algorithm::Asymmetric(alg) = self.object_info.algorithm {
+            if C::asymmetric_algorithm() == alg {
+                let mut reader = SliceReader(&self.data);
+
+                SecretKey::<C>::from_slice(reader.read(FieldBytesSize::<C>::USIZE)?).ok()
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn rsa(&self) -> Option<RsaPrivateKey> {
+        let (component_size, modulus_size) = match self.object_info.algorithm {
+            algorithm::Algorithm::Asymmetric(asymmetric::Algorithm::Rsa2048) => (128, 256),
+            algorithm::Algorithm::Asymmetric(asymmetric::Algorithm::Rsa3072) => (192, 384),
+            algorithm::Algorithm::Asymmetric(asymmetric::Algorithm::Rsa4096) => (256, 512),
+            _ => return None,
+        };
+
+        let mut reader = SliceReader(&self.data);
+
+        let p = BigUint::from_bytes_be(reader.read(component_size)?);
+        let q = BigUint::from_bytes_be(reader.read(component_size)?);
+        let dp = BigUint::from_bytes_be(reader.read(component_size)?);
+        let dq = BigUint::from_bytes_be(reader.read(component_size)?);
+        let _qinv = BigUint::from_bytes_be(reader.read(component_size)?);
+        let n = BigUint::from_bytes_be(reader.read(modulus_size)?);
+        const EXP: u64 = 65537;
+        let e = BigUint::from_u64(EXP).expect("invalid static exponent");
+
+        let d = e
+            .clone()
+            .mod_inverse((dp - BigUint::one()) * (dq - BigUint::one()))?
+            .to_biguint()?;
+
+        let private_key = RsaPrivateKey::from_components(n, e, d, vec![p, q]).ok()?;
```

Comment:
Perhaps you could open a PR for this for the `rsa` crate?

Reply:
I guess so. Any suggestion on the naming? Because the serialization is not a standard I've heard of.

`to_precomputed`/`from_precomputed`?

Reply:
Just realized this was not the only serialization format for rsa.

So they have another one for put_asymmetric_key commands: https://github.com/baloo/yubihsm.rs/commit/749c62628ebf10dc54a936458e1f8b8ff087022a#diff-1ae6dd82cd286c17a8c1c405c8b551bf8075cb56d12033688b1dec4f17ac4099R64
(a very dirty commit)

Reply:
Never mind then